### PR TITLE
gosrc2cpg: structure support

### PR DESCRIPTION
**Pull request by hiteshbedre**
Changes:
1. Added support for Go's structure.
2. Covered test cases having simple structure definition to complex(one structure used under another)

### CODE REVIEW

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForExpressionCreator.scala [hunk 1]
```diff
@@ -23,24 +23,14 @@ trait AstForExpressionCreator(implicit withSchemaValidation: ValidationMode) { t
...
-  private def astForField(field: ParserNodeInfo): Ast = {
+  protected def astForField(field: ParserNodeInfo): Ast = {
     field.node match
       case Field => {
-        Ast()
+        val typeInfo = createParserNodeInfo(field.json(ParserKeys.Type))
+        val (typeFullName, typeFullNameForcode, isVariadic, evaluationStrategy) = processTypeInfo(typeInfo)
+        val fieldNodeInfo = createParserNodeInfo(field.json(ParserKeys.Names).arr.head)
+        val fieldName     = fieldNodeInfo.json(ParserKeys.Name).str
+        Ast(memberNode(typeInfo, fieldName, s"${fieldName} ${typeFullNameForcode}", typeFullName, Seq()))
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForExpressionCreator.scala:33]:
```suggestion
        Ast(memberNode(typeInfo, fieldName, s"$fieldName $typeFullNameForCode", typeFullName, Seq()))
```

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForExpressionCreator.scala [hunk 2]
```diff
@@ -23,24 +23,14 @@ trait AstForExpressionCreator(implicit withSchemaValidation: ValidationMode) { t
...
-    Seq(Ast())
-  }
-
-  private def astForField(field: ParserNodeInfo): Ast = {
+  protected def astForField(field: ParserNodeInfo): Ast = {
     field.node match
       case Field => {
-        Ast()
+        val typeInfo = createParserNodeInfo(field.json(ParserKeys.Type))
+        val (typeFullName, typeFullNameForcode, isVariadic, evaluationStrategy) = processTypeInfo(typeInfo)
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForExpressionCreator.scala:30]:
```suggestion
        val (typeFullName, typeFullNameForCode, isVariadic, evaluationStrategy) = processTypeInfo(typeInfo)
```

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForStructuresCreator.scala [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+trait AstForStructuresCreator(implicit withSchemaValidation: ValidationMode) { this: AstCreator =>
+
+  def astForTypeSpec(typeSpecNode: ParserNodeInfo): Seq[Ast] = {
+    // TODO: Add support for member variables and methods
+    Option(typeSpecNode) match {
+      case Some(typeSpec) =>
+        val nameNode          = typeSpec.json(ParserKeys.Name)
+        val typeNode          = createParserNodeInfo(typeSpec.json(ParserKeys.Type))
+        val a: NewTypeDecl    = methodAstParentStack.collectFirst { case t: NewTypeDecl => t }.get
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForStructuresCreator.scala:24]: Could we make this
```scala
methodAstParentStack.collectFirst { case t: NewTypeDecl => t } match
  Some(a) => ...
  None => Seq.empty
```

**Review by DavidBakerEffendi — CHANGES_REQUESTED** [review]

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForExpressionCreator.scala (hunk 1 above)
Comment (hiteshbedre) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForExpressionCreator.scala:33]: Addressed  here: https://github.com/joernio/joern/pull/3567/commits/7939c8bca109bad436a02b51d50e84e12403ce1d

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForExpressionCreator.scala (hunk 2 above)
Comment (hiteshbedre) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForExpressionCreator.scala:30]: Addressed here: https://github.com/joernio/joern/pull/3567/commits/7939c8bca109bad436a02b51d50e84e12403ce1d

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForStructuresCreator.scala (hunk 3 above)
Comment (hiteshbedre) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForStructuresCreator.scala]: Addressed here: https://github.com/joernio/joern/pull/3567/commits/7939c8bca109bad436a02b51d50e84e12403ce1d

**Comment by hiteshbedre** [comment]
@DavidBakerEffendi Addressed code review comment in: https://github.com/joernio/joern/pull/3567/commits/7939c8bca109bad436a02b51d50e84e12403ce1d

**Comment by hiteshbedre** [comment]
New addressed changes:
1. Updated existing test cases in more structured way(using `in` clause)
2. Added test cases scenarios where we are moving member node to typeDecl node.

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForStructuresCreator.scala [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+trait AstForStructuresCreator(implicit withSchemaValidation: ValidationMode) { this: AstCreator =>
+
+  def astForTypeSpec(typeSpecNode: ParserNodeInfo): Seq[Ast] = {
+    // TODO: Add support for member variables and methods
+    Option(typeSpecNode) match {
+      case Some(typeSpec) =>
+        val nameNode = typeSpec.json(ParserKeys.Name)
+        val typeNode = createParserNodeInfo(typeSpec.json(ParserKeys.Type))
+        val a = methodAstParentStack.collectFirst { case t: NewTypeDecl => t } match {
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForStructuresCreator.scala:24]: I meant the whole block below should be in the `Some` part. e.g. 
```scala
methodAstParentStack.collectFirst { case t: NewTypeDecl => t } match {
   Some(a) => 
        val astParentType     = a.label
         val astParentFullName = a.fullName
         val fullName          = nameNode(ParserKeys.Name).str // TODO: Discuss fullName structure
         val typeDeclNode_ =
           typeDeclNode(
             typeSpec,
             nameNode(ParserKeys.Name).str,
             fullName,
             parserResult.filename,
             typeSpec.code,
             astParentType,
             astParentFullName
           )

         methodAstParentStack.push(typeDeclNode_)
         scope.pushNewScope(typeDeclNode_)
         val memberAsts = astForStructType(typeNode)

         methodAstParentStack.pop()
         scope.popScope()

         val modifier = addModifier(typeDeclNode_, nameNode(ParserKeys.Name).str)
         Seq(Ast(typeDeclNode_).withChild(Ast(modifier)).withChildren(memberAsts.toSeq))
    None => Seq.empty
}

Comment (hiteshbedre) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForStructuresCreator.scala]: Addressed here: https://github.com/joernio/joern/pull/3567/commits/e7443b7c48e81e5924389917f0017b0a50503e55

**Review by DavidBakerEffendi — APPROVED** [review]

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForExpressionCreator.scala [hunk 5]
```diff
@@ -23,24 +23,14 @@ trait AstForExpressionCreator(implicit withSchemaValidation: ValidationMode) { t
     }
   }
 
-  private def astForStructType(expr: ParserNodeInfo): Seq[Ast] = {
```
Comment (pandurangpatil) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForExpressionCreator.scala:26]: @hiteshbedre We can move this method as well to as this related to TypeDecl `AstForStructuresCreator.scala`

**Pull request merged**